orttraining/orttraining/training_ops/cuda/optimizer/clip_grad_norm/clip_grad_norm_impl.cu

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <algorithm>
+#include "core/providers/cuda/cuda_common.h"
+#include "orttraining/training_ops/cuda/optimizer/clip_grad_norm/clip_grad_norm_impl.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T>
+__global__ void ClipGradNorm(
+    ChunkGroup<ClipGradNormGroupSize> chunks,
+    const float* l2_norm,
+    const float epsilon,
+    const float max_norm) {
+  const int tensor_idx = chunks.block_index_to_tensor_group_index[blockIdx.x];
+  const int tensor_size = chunks.tensor_sizes[tensor_idx];
+
+  const int chunk_start_idx = chunks.block_index_to_chunk_start_index[blockIdx.x];
+  // chunk_size is chunks.chunk_size if the loaded chunk is full. Otherwise (this
+  // chunk is the last one in the source tensor), the actual size is determined
+  // by the bound of the source tensor.
+  const int chunk_size = min(tensor_size, chunk_start_idx + chunks.chunk_size) - chunk_start_idx;
+
+  T* gradients_chunk_ptr = static_cast<T*>(chunks.tensor_ptrs[0][tensor_idx]) + chunk_start_idx;
+
+#pragma unroll(4)
+  for (int i = threadIdx.x; i < chunk_size; i += blockDim.x) {
+    float clip_coefficient = max_norm / (*l2_norm + epsilon);
+    gradients_chunk_ptr[i] = static_cast<T>(gradients_chunk_ptr[i]) *
+                             static_cast<T>(fminf(clip_coefficient, 1.0f));
```

Comment:
if you have the norm_type defined, shall we make the computation logic as separate function/functor. 

Then, adding a new norm_type should just add a new branch and new functor for its computation.

Reply:
I am not sure I followed the suggestion. The norm currently is calculated in a separate kernel (for `fro`, the ReduceAllL2 kernel followed by the sqrt calculates the norm). So, the norm computation happens outside of this kernel anyway. If we plan on adding support for more norms, then we should just change the kernel being invoked for the norm calculation outside of this kernel before invoking it.